### ISSUE TITLE
[Templating] If function not recognized for ColumnSet style

### ISSUE THREAD

**Issue by pschaeflein** [microsoft/AdaptiveCards#3738]
Attempting to render a "table" with alternative row backgrounds.

Designer:
![image](https://user-images.githubusercontent.com/8008013/72924977-86015d80-3d17-11ea-9e7b-65f5448578f5.png)

CardPayload:
```json
{
    "type": "AdaptiveCard",
    "body": [
        {
            "type": "ColumnSet",
            "columns": [
                {
                    "type": "Column",
                    "width": 2,
                    "items": [
                        {
                            "type": "TextBlock",
                            "text": "Title",
                            "weight": "Bolder"
                        }
                    ]
                },
                {
                    "type": "Column",
                    "width": 1,
                    "items": [
                        {
                            "type": "TextBlock",
                            "text": "Version"
                        }
                    ]
                }
            ]
        },
        {
            "type": "ColumnSet",
            "$data": "{packages}",
            "style": "{if(($index % 2 == 0), 'default', 'emphasis')}",
            "columns": [
                {
                    "type": "Column",
                    "width": 2,
                    "items": [
                        {
                            "type": "TextBlock",
                            "text": "{title}",
                            "wrap": true
                        }
                    ]
                },
                {
                    "type": "Column",
                    "width": 1,
                    "items": [
                        {
                            "type": "TextBlock",
                            "text": "{version}"
                        }
                    ]
                }
            ]
            
        }
    ],
    "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
    "version": "1.0"
}
```
Sample data:
```json
{
    "packages": [
        {
            "title":"Product 1",
            "version":"5.1.3"
        },
        {
            "title":"Product 2",
            "version":"5.1.17"
        }
    ]
}
```

**Comment by matthidinger** [comment]
Hi Paul, thanks for taking the time to use templating and let us know your experience so far. 

Couple things:

1. There's a syntax error in your `if` function, it should be: `"style": "{if($index % 2 == 0, 'default', 'emphasis')}"`
2. Unfortunately, even with that fix it still doesn't work, because the modulo `%` operator isn't implemented yet, and...
3. You seem to have identified a bug when a property binding is applied to the `style` property. @dclaux can you confirm this?

Also @dclaux, at least on the published designer, when I try putting this `style` on a `ColumnSet`: `"style": "{if(1 == 1, 'default', 'emphasis')}"` I see the error saying it's an invalid style, but there is a data loss bug if I click `Preview Mode` it deletes my style. Do you want me to file a separate bug to track this?

**Comment by dclaux** [comment]
@matthidinger you beat me to this.
- Indeed modulus isn't currently supported (along with a bunch of other functions)
- The bug you are referring to is already fixed in vnext.adaptivecards.io, I do not plan to backport it. Also, the vnext site should be updated with the latest from js/renderer-refactor
- Beyond these two, there seems to be a problem with $index used in the if function anyway:

This: `"style": "{if($index == 1, 'default', 'emphasis')}"`

Results in this:
![image](https://user-images.githubusercontent.com/1334689/72927377-eccc4a00-3d0a-11ea-8272-4cc0213a176b.png)

which is obviously incorrect. Note I updated the TextBlock's text to be `"text": "{title} + {$index}"`, so $index functions as expected in the text property of a TextBlock. 

I will look into these as time permits.

**Comment by dclaux** [comment]
I have pushed [a change to the js/refactor branch](https://github.com/microsoft/AdaptiveCards/commit/3a0e720536498a419dc0d2e9dce928e2cea9dd32) that adds support for modulus (`%`) and I took the opportunity to also add `round`, `ceil` and `floor` functions. It also fixes the value of `$index` not being properly evaluated in a function, as well as a couple of serialization bugs that made the designer misbehave.

@pschaeflein these changes will make your templated card "work", but:
- You won't see it work in the designer because there is a bug that causes the `"style"` property to be lost when its value isn't a boolean. I haven't had time to fix that yet.
- The rendered result will not look good, and there is no good solution for that. This is caused by the way we're trying to be smart with paddings when container styles are applied. You'll see that your alternating rows are not aligned

**Comment by pschaeflein** [comment]
@dclaux Thanks David! I'll try it out when I have a free moment.

I do hope that the templating capability will eventually allow for "conditional" formatting / visibility of all elements in the card.

**Comment by dclaux** [comment]
@pschaeflein note that to try it out, for now at least, you'd have to build the library from source from the **js/renderer-refactor** branch. But @matthidinger maybe you could publish beta packages for both renderer and template engine?

@pschaeflein can you elaborate on what you mean by conditional formatting/visibility? There is already quite a  bit you can do - for instance do you know about `$when`?

**Comment by pschaeflein** [comment]
Yes,  understood about building directly.

Yes, I've seen `$when`. (I forgot about that when typing "visibility".) My point is a reaction to your comment about the current rendered result alignment. I should have been more specific -- I can see many instances where attributes of a node are set as a result of `if()`, and I would want the alignment/styling to match a card without those conditions.

**Comment by matthidinger** [comment]
This should be fixed in the latest v1.0 release, please reopen if that isn't the case
